src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/NullAnalysis/NullAnalysis.NullDataFlowOperationWalker.cs

```diff
@@ -0,0 +1,319 @@
+                {
+                    return DefaultValue;
+                }
+            }
+
+            protected override void ResetCurrentAnalysisData(NullAnalysisData newAnalysisDataOpt = null) =>
+                CurrentAnalysisData.Reset(newAnalysisDataOpt);
+
+            #region Visitor methods
+            public override NullAbstractValue DefaultVisit(IOperation operation, object argument)
+            {
+                var value = base.DefaultVisit(operation, argument);
+                if (operation.ConstantValue.HasValue)
+                {
+                    return operation.ConstantValue.Value == null ?
+                        NullAbstractValue.Null :
+                        NullAbstractValue.NotNull;
+                }
+                else if (operation.Type != null && operation.Type.IsValueType)
+                {
+                    return NullAbstractValue.NotNull;
+                }
+
+                return value;
+            }
+
+            protected override NullAbstractValue VisitAssignmentOperation(IAssignmentOperation operation, object argument)
+            {
+                var value = base.VisitAssignmentOperation(operation, argument);
+
+                if (operation.Target.Type.IsValueType)
+                {
+                    return NullAbstractValue.NotNull;
+                }
+
+                return value;
+            }
+
+            public override NullAbstractValue VisitCoalesce(ICoalesceOperation operation, object argument)
+            {
+                var leftValue = Visit(operation.Value, argument);
+                var rightValue = Visit(operation.WhenNull, argument);
+                switch (leftValue)
+                {
+                    case NullAbstractValue.NotNull:
+                        return NullAbstractValue.NotNull;
+
+                    case NullAbstractValue.Null:
+                        return rightValue;
+
+                    default:
+                        return NullAbstractValue.MaybeNull;
+                }
+            }
+
+            public override NullAbstractValue VisitConditionalAccess(IConditionalAccessOperation operation, object argument)
+            {
+                var leftValue = Visit(operation.Operation, argument);
+                var rightValue = Visit(operation.WhenNotNull, argument);
+                switch (leftValue)
+                {
+                    case NullAbstractValue.Null:
+                        return GetAbstractDefaultValue(operation.WhenNotNull.Type);
+
+                    case NullAbstractValue.NotNull:
+                        return rightValue;
+
+                    default:
+                        return NullAbstractValue.MaybeNull;
+                }
+            }
+
+            public override NullAbstractValue VisitDeclarationExpression(IDeclarationExpressionOperation operation, object argument)
+            {
+                var unused = base.VisitDeclarationExpression(operation, argument);
```

Comment:
>var unused = base.VisitDeclarationExpression(operation, argument); [](start = 16, length = 66)

Do we have a state in the base class we affect with those calls? The semantic is confusing.

Reply:
Base class is storing the value for every visited operation in a map (see Visit override in the base class). This map is eventually added to the final analysis result. If a subtype of the visitor skips visiting some nodes, then it will lead to missing entries in this map. It relies on standard visitor implementation pattern, that every single node in the tree will be visited.

However, I agree that we must add a debug check or assert to verify this contract, otherwise it can easily be violated in a subtype. I will do so in a separate commit.

Reply:
Maybe you can use discards here, or omit the assignment entirely?

Reply:
I have added debug code in `DataflowOperationVisitor.Flow` that will enforce that we Visit every single descendant operation.

---
In reply to: [162760854](https://github.com/dotnet/roslyn-analyzers/pull/1541#discussion_r162760854) [](ancestors = 162760854)